sui_programmability/framework/sources/Governance/SuiSystem.move

```diff
@@ -0,0 +1,162 @@
+        min_validator_stake: u64,
+
+        /// Upper-bound on the amount of stake allowed to become a validator.
+        max_validator_stake: u64,
+    }
+
+    /// The to-level object containing all information of the Sui system.
+    struct SuiSystemState has key {
+        id: VersionedID,
+        /// The current epoch ID, starting from 0.
+        epoch: u64,
+        /// Contains all information about the validators.
+        validators: ValidatorSet,
+        /// The SUI treasury capability needed to mint SUI.
+        treasury_cap: TreasuryCap<SUI>,
+        /// The storage fund.
+        storage_fund: Coin<SUI>,
+        /// A list of system config parameters.
+        parameters: SystemParameters,
+    }
+
+    /// Create a new SuiSystemState object and make it shared.
+    /// This function will be called only once in Genesis.
+    public(friend) fun create(
+        validators: vector<Validator>,
+        treasury_cap: TreasuryCap<SUI>,
+        storage_fund: Coin<SUI>,
+        min_validator_stake: u64,
+        max_validator_stake: u64,
+        ctx: &mut TxContext,
+    ) {
+        let state = SuiSystemState {
+            id: TxContext::new_id(ctx),
+            epoch: 0,
+            validators: ValidatorSet::new(validators),
+            treasury_cap,
+            storage_fund,
+            parameters: SystemParameters {
+                min_validator_stake,
+                max_validator_stake,
+            },
+        };
+        Transfer::share_object(state);
+    }
+
+    /// Can be called by anyone who wishes to become a validator in the next epoch.
+    /// At the end of the current epoch, the system will look at the amount of stake
+    /// compared to other validator candidates to decide if it is eligible.
+    /// If not, the `validator` object will be returned to `validator.sui_address` at that time.
+    ///
+    /// The `validator` object needs to be created before calling this.
+    /// The amount of stake in the `validator` object must meet the requirements.
+    public(script) fun request_add_validator(
+        self: &mut SuiSystemState,
+        validator: Validator,
+        _ctx: &mut TxContext,
+    ) {
+        let stake_amount = Validator::get_stake_amount(&validator);
+        assert!(
+            stake_amount >= self.parameters.min_validator_stake
+                && stake_amount <= self.parameters.max_validator_stake,
+            EINVALID_STAKE_AMOUNT
+        );
+        ValidatorSet::request_add_validator(&mut self.validators, validator);
+    }
+
+    /// A validator can call this function to request a withdraw in the next epoch.
+    /// We use the sender of `ctx` to look up the validator
+    /// (i.e. sender must match the sui_address in the validator).
+    /// At the end of the epoch, the `validator` object will be returned to the sui_address
+    /// of the validator.
+    public(script) fun request_withdraw_validator(
```

Comment:
Nit: I would call this something like `request_remove_validator`. "Withdraw" sounds like it might be about withdrawing stake rather than removing a validator from the validator set.